### DISCUSSION TITLE
Render html and next Image from Server Side Rendering

### DISCUSSION THREAD

**Discussion by 2gud4u** [vercel/next.js#38309] [Help] [answered]
Hello community !!! I want to display a header with some text and a Image (obtaining its src) all with Server Side Rendering this is what i have tried, the problem i think, because of the errors is that i dont display very well the info from the api routes i have even used swr but no luck. Could someone explain me how to acheive this and render each one at different times ? Currently i want to use a timer for that with a countdown time
Heres some of my code.-

Api routes

in /api/education theres a index.js with this code.-

```
export default function handler(req, res) {
  // const jsonData = await getData()
  res.status(200).json({name: 'to your new Home for personalized education'})
}
```

in /api/logo theres a index.js with this code.-

```
export default function handler(req, res) {
  // const jsonData = await getData()
  res.status(200).json({src:'http://localhost:3000/logo/EdAIazul.gif'})
}
```

in the component i make the api calls in the return function i typed something like this.-

```
import Image from 'next/image';

return(
	<h1>{dataeducation.name}<h1>
	<Image src={datalogo.src} alt=’Image logo’/>
)

```
this is my getServerSideProps in the same component (as it calls to api routes i dont know if it is correct.-

```
export async function getServerSideProps() {
    // const res = await fetch('https://');
    // const newRender = await res.json();

    const education = await fetch('http://localhost:3000/api/education.js')
    const educationData = await education.json()
    const logo = await fetch('http://localhost:3000/api/logo.js')
    const logoData = await logo.json()

    return {
        props: {
            educationData,
            logoData
        }
    }
}
```

Im using “next”:"^12.2.0",
"react": "18.2.0",
"react-dom": "18.2.0",

**Comment by icyJoseph** [comment, marked as answer]
Hi,

I am not sure I follow how you want to design this, but a couple of pointers:

- `getServerSideProps` is already a sort of specialised API route, that can pass its result to a Page component
- Calling api routes from `getServerSideProps` _is an unnecessary jump_, you can do any kind of data fetching you'd in api routes inside `getServerSideProps` as well.
- Because API routes are URL endpoints you don't need to add the `.js`, this applies also when calling them from the client

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

The above is from the [next docs](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes).

---

In your case, I imagine, for example:

```ts
// this function should be in a folder `lib` for example, `lib/education`, from where you can import it to this page
// and to `api/education` in case you wanted to have the option to reach `api/education` from other means
const fetchEducationData = () => fetch("https://education.data.api.com/");

export async function getServerSideProps() {
    const education = await fetchEducationData();
    const educationData = await education.json()

   // this would have to return the first logo you want to show, so you can directly do:
    const logoData = {src: 'path/to-the-first/image'};

    return {
        props: {
            educationData,
            logoData
        }
    }
}
```

Notice, I've refactored `fetchEducationData`  into something that you could also use in `api/education` as well. 

Now from your client, you can:

- Initialise state to the value of the first image you got from `getServerSideProps`
- inside a `useEffect`, put an timeout or interval that calls `api/logo` like `fetch('/api/logo')`, and gets the next image src. 
- Once you have the next image src, set state to that value
- Pass that state to the Image component

I assume for some reason you cannot just hard code the list of possible images on your component and cycle through it. It may come from a CMS or stuff like that.

Perhaps you just want to learn and try things out and that's very good. If you have any follow up questions or a repository to look at I can help out.

**Reply by 2gud4u** [reply]
Right i just want to learn, and as i read getServerSideProps for Server Rendering is the best, i wanted to give it a try, then i read also about SWR, and some other things, your answer help me a lot @icyJoseph , in fact i think it would fix my issue, i wont mark it as answer, because i want this post to be a collection of ideas, and if someone sees this, he/she can do whatever they want in Next.js. Thank you very much !!!  EDIT: Oh just one thing in the call to the api of useEffect for the logo how do i pass that value tu the return of the component (hope i explained correctly) Oh yes through the state right Thanks !!!

**Reply by 2gud4u** [reply]
Sorry @icyJoseph , it seems to work with timers and all but as i have already rendered content all that must be overwritten, with that i mean the timers and apis content has to replace another content already displayed, could that be the cause of my errors now?
